### ISSUE TITLE
Vine: Insufficient Error Message in Serverless

### ISSUE THREAD

**Issue by JinZhou5042** [cooperative-computing-lab/cctools#3570]
In serverless computing, specifying incorrect package names in the `import_modules` parameter halts the program and triggers an error. For instance, with import_modules = ["osa"], the system generates the following error:

```
ModuleNotFoundError: No module named 'osa'
```

Conversely, when a submodule of a correctly named supermodule is incorrectly specified, the program does not stop or produce an error message. This behavior was observed under the following conditions:

```python
import_modules = ["os.pat"]
import_modules = [("os.pat", "path")]
```
In both cases, no errors are reported, and the program appears to run successfully, yet it fails to complete any tasks.

![Screenshot from 2023-11-13 20-08-24](https://github.com/cooperative-computing-lab/cctools/assets/142265839/611c51a5-bcd2-48c1-a482-5869001ecfd2)

I suspect this issue arises because the worker fails but does not report its error to the manager.  

Should we consider addressing this issue?

**Comment by btovar** [comment]
This is the reason I didn't want to use strings. Since import_modules is going to be used only for modules, and the function themselves will retain the import statements, perhaps you can use the modules directly and remove the `import x as y` part. In this way, if a module is incorrectly specified, we will know at the time the library is created. E.g.,

```python
..., import_modules = [ os.path ], ...

def my_fun(...):
    import os.path
    ....
```

**Comment by JinZhou5042** [comment]
That makes sense to me. What I am concerned about is for some well-known modules, they have conventional aliases (such as `np` for `numpy`, `pd` for `pandas`), it will be still time-consuming by delaying the `import x as y` statement from the preamble to the function itself, thus I think it's better to keep this feature.

Plus, the `__name__` attribute of some module (e.g., `os.path`) is depending on the running platform (`posixpath` for Unix, `ntpath` for Windows), since the user might be unaware of this nuanced usage, its better to explicitly indicate the name of packages used in functions for every potentially relevant ones.

Based on the concern, I recommend using a combination of the module itself and a string name that is actually used in functions. E.g.,

```python
import numpy as np
import pandas as pd
import math
import os.path

import_modules = [(numpy, 'np'), (pandas, 'pd'), (math, 'math'), (os.path, 'os.path')]
```

The argument will be translated into:

```python
import numpy as np
import pandas as pd
import math as math
import posixpath as os.path
```

I think by doing this we can ensure packages can be correctly imported, in spite of the fact the user might think we give him a tricky usage. How do you think @btovar

**Comment by JinZhou5042** [comment]
Additionally, it is still the user's responsibility to make sure they body of their functions were correctly written. For example, the following function will not exit our program and keep running but with no tasks completed.
```python
def cube(x):
    from random import uniform1
    from time import sleep as time_sleep
    ansdioanoi()        
    asdas
    sarandom_delsay = unifoadrm(0.00001, 0.0001)
    timeasd_sleep(randomasd_dyhelay)
```
This is how I print output information:
```python
while not q.empty():
        t = q.wait(5)
        print(t)
        if t:
            x = t.output
            total_sum += x
            print(f"task {t.id} completed with result {x}")
```
It results in the following display:
![Screenshot from 2023-11-14 09-35-22](https://github.com/cooperative-computing-lab/cctools/assets/142265839/fc491075-2645-42b5-8f28-b7e49ab624ba)

**Comment by JinZhou5042** [comment]
One weird behavior was observed:

For `/cctools/taskvine/src/examples/vine_example_function_call.py`, if we set physical resources for it this way:
```python
libtask = q.create_library_from_functions('test-library', divide, double, cube, import_modules=import_modules)

libtask.set_cores(1)
libtask.set_memory(1000)
libtask.set_disk(1000)

q.install_library(libtask)
```
It results in keeping running with no results returned once again. 

But for `/cctools/taskvine/test/vine_python_serverless.py`, most of the code in it is the same as the previous one (what different is this script is launched by `/cctools/taskvine/test/TR_vine_python_serverless.sh`), but setting cores, memory and disk just results in successful running.

**Comment by btovar** [comment]
Let's check, but I don't think that adding a name to a module adds noticeable overhead. E.g.:

```python
import numpy

def myfun():
    import numpy as np
    ....
````

Further, let's check that:

```
import posixpath

def myfun():
    import os.path as p
    ....
````

also does not lead to overhead. E.g.,  the library may load posixpath directly, and when the function tries to import `os.path`, it won't load `posixpath` again.

**Comment by btovar** [comment]
Let's avoid strings and module aliases names (e.g. (numpy, "np")) if possible.

**Comment by JinZhou5042** [comment]
This is doable, let me run the test at scale.

**Comment by dthain** [comment]
The underlying problem here is not really even Python, but that fact that we don't have an error return channel for LibraryTask instances that fail.

**Comment by dthain** [comment]
So, per our discussion today. The problem here is not about imports specifically, but rather about what to do when a library files.

The short-term solution we agreed upon is this:
- Our construction of library tasks should capture stdout/err in a file that can be brought back.
- When a library task fails, the manager should printf the stdout/err of the library task, which should come back as a normal file or buffer.
- Keep track of the number failures in the prototype library object.
- After a some number of failures, stop printing things out.

**Comment by dthain** [comment]
@JinZhou5042 please use this issue to write up how library/function dispatch currently works, and what the current limitations are.

**Comment by dthain** [comment]
Then, once we have a clear understanding, propose some solutions for us to discuss.  Then, once we have some agreement, start writing code.

**Comment by dthain** [comment]
Fixed by #3855
